typegate/src/typegate/artifacts/shared.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+
+import { createHash } from "node:crypto";
+import { S3, S3ClientConfig } from "aws-sdk/client-s3";
+import {
+  deinitUploadUrlStore,
+  getLocalPath,
+  initUploadUrlStore,
+  STORE_DIR,
+  STORE_TEMP_DIR,
+  UploadUrlStore,
+} from "./local.ts";
+import { ArtifactStore } from "./mod.ts";
+import { HashTransformStream } from "../../utils/hash.ts";
+import { resolve } from "std/path/resolve.ts";
+
+export class SharedArtifactStore extends ArtifactStore {
+  #uploadUrls: UploadUrlStore;
+  #s3!: S3;
+  #s3Bucket: string;
+
+  static init(s3Config: S3ClientConfig, s3Bucket: string) {
+    return new SharedArtifactStore(s3Config, s3Bucket);
+  }
+
+  constructor(s3Config: S3ClientConfig, s3Bucket: string) {
+    super();
+    this.#s3 = new S3(s3Config);
+    this.#s3Bucket = s3Bucket;
+    this.#uploadUrls = initUploadUrlStore();
```

Comment:
The upload url store in this case should be stored in Redis...

Reply:
oww, ok. Makes sense, I will change to that.

Reply:
Done.